### PR TITLE
Settings: add FPS cap

### LINKED ISSUE #487: Allow setting Max FPS value

### PR DESCRIPTION
**Pull request by archbirdplus**
Adds `fpsCap` field which lets Cubyz sleep if it processes the frame fast enough. On lower end hardware, this would let Cubyz deliver a more consistent framerate, albeit lower.

- [x] settings.json option
- [x] in-game menu slider
- [x] update physics to use real time, not rendering time
- [ ] ~~update fps label in debug window to count actual frames, not inverse rendering time~~
- [x] update fps label to use real time between frames, not rendering time

Resolves #487.

### CODE REVIEW

**Comment by archbirdplus** [comment]
Actually, I would group point 4 along with #438.

File: src/main.zig [hunk 1]
```diff
@@ -446,17 +449,27 @@ pub fn main() void {
...
+		}
+		lastFrameTime.store(frameTime, .monotonic);
+
+		if(settings.fpsCap) |fpsCap| {
+			const minFrameTime = @divFloor(1000*1000*1000, fpsCap);
+			const sleep = @min(minFrameTime, @max(0, minFrameTime - (endRendering -% lastBeginRendering)));
+			std.time.sleep(sleep);
+		}
+		const begin = std.time.nanoTimestamp();
+		const deltaTime = @as(f64, @floatFromInt(begin - lastBeginRendering))/1e9;
```
Comment (IntegratedQuantum) [inline src/main.zig:466]: begin -% lastBeginRendering, otherwise it crashes once in a century or something like that.

**Comment by archbirdplus** [comment]
Sliders are bugged for me in general. We might want to round the fps from the slider because people probably usually just want to reset it to 60 fps without a hassle. On the other hand, some people prefer to set their framerate just _under_ the refresh rate for some reason, so I'm not sure which is a priority.

**Comment by IntegratedQuantum** [comment]
You could also make it a text field. Then players could be arbitrarily precise.

File: src/gui/windows/graphics.zig [hunk 2]
```diff
@@ -24,6 +24,14 @@ const anisotropy = [_]u8{1, 2, 4, 8, 16};
 
 const resolutions = [_]u16{25, 50, 100};
 
+fn fpsCapFormatter(allocator: main.utils.NeverFailingAllocator, value: f32) []const u8 {
+	return std.fmt.allocPrint(allocator.allocator, "#ffffffFPS Limit: {d:.0}", .{value}) catch unreachable;
+}
+
+fn fpsCapCallback(newValue: f32) void {
+	settings.fpsCap = if(newValue == 144.0) null else @intFromFloat(newValue);
```
Comment (IntegratedQuantum) [inline src/gui/windows/graphics.zig:32]: What about 60 Hz monitors? Is there a way to get the monitor refresh rate from glfw maybe?

Comment (archbirdplus) [reply src/gui/windows/graphics.zig:32]: 60 Hz is unaffected by this line. Of course, now I realize that it prevents people from using all 144 frames of their 144 monitors :(. I think I'll just round the slider to the nearest 5, snap to 144, and unlimited.

File: src/gui/windows/graphics.zig [hunk 3]
```diff
@@ -24,12 +24,25 @@ const anisotropy = [_]u8{1, 2, 4, 8, 16};
...
+	} else {
+		return null;
+	}
+}
+
 fn fpsCapFormatter(allocator: main.utils.NeverFailingAllocator, value: f32) []const u8 {
-	return std.fmt.allocPrint(allocator.allocator, "#ffffffFPS Limit: {d:.0}", .{value}) catch unreachable;
+	const cap = fpsCapRound(value);
+	const limit = if(cap == null) "Unlimited" else std.fmt.allocPrint(allocator.allocator, "{d:.0}", .{cap.?}) catch unreachable;
+	return std.fmt.allocPrint(allocator.allocator, "#ffffffFPS: {s}", .{limit}) catch unreachable;
```
Comment (IntegratedQuantum) [inline src/gui/windows/graphics.zig:40]: Shouldn't this still be `FPS Limit`? Having just fps might be misleading.

File: src/gui/windows/debug.zig [hunk 4]
```diff
@@ -25,7 +25,7 @@ pub var window = GuiWindow {
 pub fn render() void {
 	draw.setColor(0xffffffff);
 	var y: f32 = 0;
-	draw.print("    fps: {d:.0} Hz{s}", .{1.0/main.lastFrameTime.load(.monotonic), if(main.settings.vsync) @as([]const u8, " (vsync)") else ""}, 0, y, 8, .left);
+	draw.print("    fps: {d:.0} Hz{s}", .{1.0/main.lastDeltaTime.load(.monotonic), if(main.settings.vsync) @as([]const u8, " (vsync)") else ""}, 0, y, 8, .left);
```
Comment (IntegratedQuantum) [inline src/gui/windows/debug.zig:28]: Could you maybe also add the fps limit in parenthesis, if the fps is limited?

Comment (archbirdplus) [reply src/gui/windows/debug.zig:28]: Changed it to either
```
a) FPS Limit: N
b) FPS: Unlimited
```

**Comment by IntegratedQuantum** [comment]
It crashes when I move the slider to the end:
```
Segmentation fault at address 0x1211879
/home/mint/Cubyz/compiler/zig/lib/compiler_rt/memset.zig:19:14: 0x17fb1d0 in memset (compiler_rt)
            d[0] = c;
             ^
/home/mint/Cubyz/compiler/zig/lib/std/mem/Allocator.zig:313:26: 0x137d134 in free__anon_31205 (Cubyzig)
    @memset(non_const_ptr[0..bytes_len], undefined);
                         ^
/home/mint/Downloads/Cubyz/src/utils.zig:749:22: 0x134c6de in free__anon_29033 (Cubyzig)
  self.allocator.free(memory);
                     ^
/home/mint/Downloads/Cubyz/src/gui/components/ContinuousSlider.zig:91:27: 0x14a39db in updateLabel (Cubyzig)
 main.globalAllocator.free(self.currentText);
                          ^
/home/mint/Downloads/Cubyz/src/gui/components/ContinuousSlider.zig:104:19: 0x14ef408 in updateValueFromButtonPos (Cubyzig)
  self.updateLabel(value, self.size[0]);
                  ^
/home/mint/Downloads/Cubyz/src/gui/components/ContinuousSlider.zig:142:32: 0x14814d4 in render (Cubyzig)
  self.updateValueFromButtonPos();
                               ^
/home/mint/Downloads/Cubyz/src/gui/gui_component.zig:102:17: 0x13f98b8 in render (Cubyzig)
     impl.render(mousePosition);
                ^
/home/mint/Downloads/Cubyz/src/gui/components/VerticalList.zig:141:15: 0x1481c9c in render (Cubyzig)
  child.render(mousePosition - shiftedPos);
              ^
/home/mint/Downloads/Cubyz/src/gui/gui_component.zig:102:17: 0x13f9930 in render (Cubyzig)
     impl.render(mousePosition);
                ^
/home/mint/Downloads/Cubyz/src/gui/GuiWindow.zig:467:19: 0x1361148 in render (Cubyzig)
  component.render((mousePosition - self.pos)/@as(Vec2f, @splat(self.scale)));
                  ^
/home/mint/Downloads/Cubyz/src/gui/gui.zig:539:16: 0x1327fd0 in updateAndRenderGui (Cubyzig)
  window.render(mousePos);
               ^
/home/mint/Downloads/Cubyz/src/main.zig:486:26: 0x131e8eb in main (Cubyzig)
   gui.updateAndRenderGui();
                         ^
/home/mint/Cubyz/compiler/zig/lib/std/start.zig:514:22: 0x131db19 in main (Cubyzig)
            root.main();
                     ^
../sysdeps/nptl/libc_start_call_main.h:58:16: 0x7f01d253ed8f in __libc_start_call_main (../sysdeps/x86/libc-start.c)
../csu/libc-start.c:392:3: 0x7f01d253ee3f in __libc_start_main_impl (../sysdeps/x86/libc-start.c)
???:?:?: 0x131d7c4 in ??? (???)
???:?:?: 0x0 in ??? (???)
```
I'm guess this happens because you return a global string from your format function, which gets freed later.

**Comment by archbirdplus** [comment]
Does it crash in release mode for you?

File: src/gui/windows/debug.zig [hunk 5]
```diff
@@ -25,7 +25,7 @@ pub var window = GuiWindow {
 pub fn render() void {
 	draw.setColor(0xffffffff);
 	var y: f32 = 0;
-	draw.print("    fps: {d:.0} Hz{s}", .{1.0/main.lastDeltaTime.load(.monotonic), if(main.settings.vsync) @as([]const u8, " (vsync)") else ""}, 0, y, 8, .left);
+	draw.print("    fps: {d:.0} Hz{s}{s}", .{1.0/main.lastDeltaTime.load(.monotonic), if(main.settings.vsync) @as([]const u8, " (vsync)") else "", if(main.settings.fpsCap == null) " (unlimited)" else std.fmt.allocPrint(main.stackAllocator.allocator, " (limit: {d:.0} Hz)", .{main.settings.fpsCap.?}) catch unreachable}, 0, y, 8, .left);
```
Comment (IntegratedQuantum) [inline src/gui/windows/debug.zig:28]: I would remove the unlimited tag. `(vsync) (unlimited)` also seems a bit contradicting.

**Comment by IntegratedQuantum** [comment]
It not crash in ReleaseFast, but it does crash in ReleaseSafe.

File: src/gui/windows/debug.zig (hunk 5 above)
Comment (IntegratedQuantum) [reply src/gui/windows/debug.zig:28]: Also I think this is leaking memory.

Comment (archbirdplus) [reply src/gui/windows/debug.zig:28]: So I suppose the 3 cases should be these:
```
(vsync)
(limit: N) (vsync)
(unlimited)
```

Comment (IntegratedQuantum) [reply src/gui/windows/debug.zig:28]: I'm would be fine with not having anything if it's unlimited.

**Comment by archbirdplus** [comment]
Zig no longer reports memory issues, but I don't think I accomplished that in the cleanest way.

File: src/gui/windows/debug.zig [hunk 6]
```diff
@@ -25,7 +25,14 @@ pub var window = GuiWindow {
 pub fn render() void {
 	draw.setColor(0xffffffff);
 	var y: f32 = 0;
-	draw.print("    fps: {d:.0} Hz{s}", .{1.0/main.lastFrameTime.load(.monotonic), if(main.settings.vsync) @as([]const u8, " (vsync)") else ""}, 0, y, 8, .left);
+	const fpsCapText = if(main.settings.fpsCap) |fpsCap| std.fmt.allocPrint(main.stackAllocator.allocator, " (limit: {d:.0} Hz)", .{fpsCap}) catch unreachable else null;
```
Comment (IntegratedQuantum) [inline src/gui/windows/debug.zig:28]: You can do `else ""`
freeing slices of length zero is explicitly allowed by the allocator interface.
Then you can just do `defer main.stackAllocator.free(fpsCapText)`

Comment (archbirdplus) [reply src/gui/windows/debug.zig:28]: Wouldn't that try to free a global string?

Comment (archbirdplus) [reply src/gui/windows/debug.zig]: I see. https://ziglang.org/documentation/master/std/#std.mem.Allocator.free. Updated, hopefully they won't change this too soon then.

Comment (IntegratedQuantum) [reply src/gui/windows/debug.zig]: Nah, they won't change it. Many data structures in the standard library already depend on it.

**Comment by IntegratedQuantum** [comment]
Could you also fix the merge conflicts?

**Comment by archbirdplus** [comment]
Fixed.

**Pull request merged**